Fix loading GDExtension dependencies on Android

m4gr3d opened the pull request:
Update the Android `gdextension` loading logic so that it copies any specified library dependencies alongside the gdextension shared library.

Fixes https://github.com/godotengine/godot/issues/87785

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

m4gr3d committed:
Fix loading GDExtension dependencies on Android

dsnopek commented on core/extension/gdextension.cpp:
```diff
@@ -686,7 +721,18 @@ GDExtensionInterfaceFunctionPtr GDExtension::get_interface_function(const String
+	return open_library_with_dependencies(p_path, p_entry_symbol, dependencies);
+}
+
+Error GDExtension::open_library_with_dependencies(const String &p_path, const String &p_entry_symbol, const Vector<SharedObject> &p_dependencies) {
```
I assume this new method was added to avoid changing the signature of `GDExtension::open_library()` because it's exposed?

This actually reminds me that we decided at a recent GDExtension meeting to un-expose this method - I'll make a PR for that soon. :-)

dsnopek replied:
Here's the PR https://github.com/godotengine/godot/pull/88418 to unbind that method

Once it's no longer exposed, I think it would be fine to add the new arguments directly on `GDExtension::open_library()`

m4gr3d replied:
Updated!

dsnopek commented on platform/android/java/lib/build.gradle:
```diff
@@ -168,7 +169,7 @@ android {
             tasks.create(name: taskName, type: Exec) {
                 executable sconsExecutableFile.absolutePath
-                args "--directory=${pathToRootDir}", "platform=android", "store_release=${storeRelease}", "production=${productionBuild}", "dev_mode=${devBuild}", "dev_build=${devBuild}", "tests=${runTests}", "target=${sconsTarget}", "arch=${selectedAbi}", "-j" + Runtime.runtime.availableProcessors()
+                args "--directory=${pathToRootDir}", "debug_symbols=${debugSymbols}", "platform=android", "store_release=${storeRelease}", "production=${productionBuild}", "dev_mode=${devBuild}", "dev_build=${devBuild}", "tests=${runTests}", "target=${sconsTarget}", "arch=${selectedAbi}", "-j" + Runtime.runtime.availableProcessors()
```
Possibly an unrelated change that snuck in?

m4gr3d replied:
Fixed!

dsnopek reviewed (commented):
Heh, just yesterday, I was just explaining to someone that the dependencies aren't used at all when loading a GDExtension, only in exporting. I guess that won't be true anymore, at least for Android. :-)

At a high-level, this looks good to me! I haven't done any testing.

m4gr3d commented:
> Discussed at the GDExtension meeting, and the `OS::open_dynamic_library()` method is really starting to acquire a lot of arguments.
> 
> (This unrelated GDExtension PR also adds a new argument: #87117)
> 
> Perhaps we should consider having a more generic `OS::open_dynamic_library()` with the minimum of argument, and then another `OS::open_gdextension_library()` that takes lots of GDExtension-specific arguments, and the OS can do what is necessary for GDExtension? It should be OK to rework this since this method isn't exposed.

What about adding a gdextension specific argument which contains all the gdextension supplemental parameters and is `nullptr` by default for regular calls.
My concern with another method is possible confusion about what to use when since at its core a gdextension is a dynamic library.

dsnopek commented:
I was imagining that the `open_dynamic_library()` would just do the low-level loading bits, and that `open_gdextension_library()` would probably call `open_dynamic_library()` (although, since both would be defined by the `OS` sub-class, they really could do it in whatever way made the most sense for the platform). It would only be GDExtension that would call `open_gdextension_library()` - any other code would call `open_dynamic_library()`.

> What about adding a gdextension specific argument which contains all the gdextension supplemental parameters and is `nullptr` by default for regular calls.

But, yes, that could work too! Are you imagining like a `struct` that has all the GDExtension info?

m4gr3d commented:
> But, yes, that could work too! Are you imagining like a struct that has all the GDExtension info?

Yeah somethink like a `GDExtensionData` similar to the pattern we have in other places like [CustomExportData](https://github.com/godotengine/godot/blob/87d40ba743ff3a3a484db051c11b16559604efa9/platform/android/export/gradle_export_util.h#L64)

dsnopek commented:
While it isn't required to solve this now (because this is a purely internal API and we can change it at any time), I'd personally prefer if we made a `struct` (as discussed above) or some other strategy to reduce the number of arguments to `OS::open_dynamic_library()`. I don't think this is the last GDExtension-related piece of data that we're going to want to pass to this method.

m4gr3d committed:
Collapse the gdextension arguments into the `GDExtensionData` struct

This is used to reduce the number of arguments to `OS::open_dynamic_library(...)`.

m4gr3d commented:
> While it isn't required to solve this now (because this is a purely internal API and we can change it at any time), I'd personally prefer if we made a `struct` (as discussed above) or some other strategy to reduce the number of arguments to `OS::open_dynamic_library()`. I don't think this is the last GDExtension-related piece of data that we're going to want to pass to this method.

@dsnopek I've added the change into a second commit since the changes are orthogonal to each other, and to make it easier to review.

Let me know if you want me to collapse the commits together.

dsnopek reviewed (approved):
Thanks! I haven't tested it, but the code looks good to me :-)

m4gr3d commented:
> Thanks! I haven't tested it, but the code looks good to me :-)

I validated the fix in https://github.com/godotengine/godot/issues/87785#issuecomment-2066760827

akien-mga commented:
Thanks!
